services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java

```diff
@@ -1033,45 +1034,40 @@ public static Response finishedRequiredActions(KeycloakSession session, Authenti
         return redirectAfterSuccessfulFlow(session, realm, userSession, clientSessionCtx, request, uriInfo, clientConnection, event, authSession);
     }
 
-    // Return null if action is not required. Or the name of the requiredAction in case it is required.
+    // Return null if action is not required. Or the alias of the requiredAction in case it is required.
     public static String nextRequiredAction(final KeycloakSession session, final AuthenticationSessionModel authSession,
-                                            final HttpRequest request, final EventBuilder event) {
-        final RealmModel realm = authSession.getRealm();
-        final UserModel user = authSession.getAuthenticatedUser();
-        final ClientModel client = authSession.getClient();
+            final HttpRequest request, final EventBuilder event) {
+        final var realm = authSession.getRealm();
+        final var user = authSession.getAuthenticatedUser();
 
         evaluateRequiredActionTriggers(session, authSession, request, event, realm, user);
 
-        Optional<String> reqAction = user.getRequiredActionsStream().findFirst();
-        if (reqAction.isPresent()) {
-            return reqAction.get();
-        }
-        if (!authSession.getRequiredActions().isEmpty()) {
-            return authSession.getRequiredActions().iterator().next();
-        }
-
-        String kcAction = authSession.getClientNote(Constants.KC_ACTION);
-        if (kcAction != null) {
-            return kcAction;
+        final var kcAction = authSession.getClientNote(Constants.KC_ACTION);
+        final var firstApplicableAction =
```

Comment:
Just a single nitpick: how about the more generic "nextApplicable"?.

Reply:
I agree, that's a better name - adjusted.